qiskit/backends/ibmq/ibmqprovider.py

```diff
@@ -39,14 +40,46 @@ def __init__(self, token, url,
     def get_backend(self, name):
         return self.backends[name]
 
-    def available_backends(self, filters=None):
+    def available_backends(self, *filters):
+        """Get a list of available backend instances from this provider.
+
+        Filters can be applied to get a subset of available backends.
+        Each filter can be of the following type:
+        dict: specifies exact-matching criteria for backend configuration
+        callable (more flexible):
+            - acceptor function: backend (IBMQBackend) -> bool
+            - selector function: backends (list(IBMQBackend) -> backends (list(IBMQBackend))
+        """
         # pylint: disable=arguments-differ
         backends = self.backends
 
-        filters = filters or {}
-        for key, value in filters.items():
-            backends = {name: instance for name, instance in backends.items()
-                        if instance.configuration.get(key) == value}
+        for filter_ in filters:
+            if isinstance(filter_, dict):
+                # exact match configuration filter:
+                # e.g. {'n_qubits': 5, 'local': False}
+                for key, value in filter_.items():
+                    backends = {name: instance for name, instance in backends.items()
+                                if instance.configuration.get(key) == value}
+            elif callable(filter_):
+                # acceptor filter: accept or reject a specific backend
+                # e.g. lambda x: x.configuration['n_qubits'] > 5
+                try:
+                    backends = {name: instance for name, instance in backends.items()
+                                if filter_(instance)}
+                except TypeError:
```

Comment:
I'm unsure about this third case - maybe because of the `available_backends(filter_backends)` example provided, which returns only one backend and feels more like an "aggregation" or "ordering" rather than a filter. What is the advantage of trying to handle that logic inside `available_backends` via.
```
filter_5 = least_busy
backends = ibmq_provider.available_backends(filter_5)
```
versus doing directly:
```
backends = imbq_provider.available_backends()
backends = least_busy(backends)
```

My concern is about trying to push a lot of functionality inside this function (which might also require some tweaking of the try..except blocks) which can already be done by directly using the backend instances and having the usual Python machinery directly. Is it mostly for solving #404? 

Can you actually add some tests for the functionality, as it might help the discussion?

Reply:
My logic for the entire PR was to make filtering more powerful. The default filter (which I realize you wrote quickly just for demonstration) was restrictive. It could only handle attribute "equality", and it only worked for `configuration`. So you wouldn't even be able to filter based on status.

So with that logic, I wanted to provide a way to write filter "functions". But philosophically, a filter can be of 2 types:
- `acceptor`: you pass each item through the filter one-by-one, and it either accepts or rejects. Each item is evaluated separately, on its own merits. This function would take Backend to bool.

- `selector`: you pass all the items to the filter, and it decides on a subset of them. This filter uses relational properties between the items, such as the `least_busy`. This function takes list(Backends) to list(Backends).
The `least_busy` is provided just as an example, and one that has been requested a lot. But a user can feed in any function.

Again as I mentioned, the reason we have filters is convenience. Otherwise your second example could be generalized: get a list of all backends, and apply your own dict/filter/whatever function over them :)

Tests, yes I'll add.

Reply:
> Again as I mentioned, the reason we have filters is convenience. Otherwise your second example could be generalized: get a list of all backends, and apply your own dict/filter/whatever function over them :)

That is actually part of my rationale for the comment - now that `available_backends` returns a list of instances, if we want flexibility, we cannot come up with something more powerful and flexible than just letting the user use that list directly and work with the instances themselves (and I would actually encourage that in the core).

I think it is about where we draw the line about convenience (vs practicality and code simplicity both in the sdk and for the users). For example, if a user wants to have a custom filter, I cannot see a situation where he would find more practical or natural to do:
```
def my_filter(backends):
   do_something_with_backends

backends = available_backends(my_filter)
```
rather than:
```
backends = my_filter(available_backends())
```

If you think there is a valid use case, that's fine! However, I think we cannot realistically enforce things like:
> This filter uses relational properties between the items, such as the least_busy

in a sane way - we are basically accepting a plain function. This opens up some questions (for example, in the current implementation, that function can only have one parameter - custom filters with extra parameters cannot use the facility; or how to deal with exceptions: now `available_backends` is open to raising other kinds of exceptions if a function filter is not working, etc): my main point is that it might be a better idea to just keep things simple and not reinvent the wheel. But again, if you strongly think there is an advantage, we can try it out (although we should at least adjust the `try..catch` to deal properly with the function, and update the docstring in the process).